crux-lucene/src/crux/lucene.clj

```diff
@@ -10,19 +10,54 @@
   (:import crux.query.VarBinding
            java.io.Closeable
            java.nio.file.Path
+           java.time.Duration
+           java.util.Date
            org.apache.lucene.analysis.Analyzer
            org.apache.lucene.analysis.standard.StandardAnalyzer
            [org.apache.lucene.document Document Field$Store StoredField StringField TextField]
            [org.apache.lucene.index DirectoryReader IndexWriter IndexWriterConfig Term]
            org.apache.lucene.queries.function.FunctionScoreQuery
            org.apache.lucene.queryparser.classic.QueryParser
            [org.apache.lucene.search BooleanClause$Occur BooleanQuery$Builder DoubleValuesSource IndexSearcher Query ScoreDoc SearcherManager TermQuery TopDocs]
-           [org.apache.lucene.store Directory FSDirectory]))
+           [org.apache.lucene.store Directory FSDirectory]
+           [java.util.concurrent Future Executors TimeUnit]))
 
-(defrecord LuceneNode [directory analyzer index-writer searcher-manager indexer]
+(defn debounce [name f wait-ms]
+  (let [!last-run-time (atom (.getTime (Date.)))
+        !is-scheduled? (atom false)
+        executor (Executors/newSingleThreadScheduledExecutor (cio/thread-factory name))
+        schedule-run! (fn run []
+                        (when (false? @!is-scheduled?)
```

Comment:
* There's a couple of potential race conditions in here - (_very_ generally speaking) any time you're reading state twice, or reading it and writing it in two distinct operations, you may have a race. `swap!` might be out given it retries, and you've also got side effects to think about - it might be that an approach based on `compare-and-set!` is your best option.
* 'reading state' here includes getting the current time :wink: time is fun.
* Worth considering what you want to happen if the fsync fails, for whatever reason
* `!is-scheduled?` is a bit of a misnomer given it might contain the job - maybe `!fut` is sufficient (i.e. anything without the `?`). fine to have an atom that's either a Future or nil.
* Looking at lodash, I think this may be 'throttling' rather than 'debouncing'?

Reply:
I've now tried to consolidate all the atom usage and current time lookups. However I already foresee that I may need to rename my three-valued `fut` :)